src/cache/cluster/cluster.go

```diff
@@ -0,0 +1,298 @@
+	}
+	list, err := memberlist.Create(c)
+	if err != nil {
+		log.Fatalf("Failed to create new memberlist: %s", err)
+	}
+	n := list.LocalNode()
+	log.Notice("Memberlist initialised, this node is %s / %s:%d", n.Name, n.Addr, port)
+	return &Cluster{
+		list:    list,
+		clients: map[string]pb.RpcServerClient{},
+	}
+}
+
+// JoinCluster joins an existing plz cache cluster.
+func (cluster *Cluster) Join(members []string) {
+	// Talk to the other nodes to request to join.
+	if _, err := cluster.list.Join(members); err != nil {
+		log.Fatalf("Failed to join cluster: %s", err)
+	}
+	for _, node := range cluster.list.Members() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		if client, err := cluster.getRPCClient(node.Addr.String() + ":" + string(node.Meta)); err != nil {
+			log.Error("Error getting RPC client for %s: %s", node.Addr, err)
+		} else if resp, err := client.Join(ctx, &pb.JoinRequest{
+			Name:    cluster.list.LocalNode().Name,
+			Address: cluster.list.LocalNode().Addr.String(),
+		}); err != nil {
+			log.Error("Error communicating with %s: %s", node.Addr, err)
+		} else if !resp.Success {
+			log.Fatalf("We have not been allowed to join the cluster :(")
+		} else {
+			cluster.hashStart = resp.HashBegin
+			cluster.hashEnd = resp.HashEnd
+			cluster.nodes = resp.Nodes
+			cluster.size = int(resp.Size)
+			return
+		}
+	}
+	log.Fatalf("Unable to contact any other cluster members")
+}
+
+// InitCluster seeds a new plz cache cluster.
+func (cluster *Cluster) Init(size int) {
+	cluster.size = size
+	// We're node 0
+	node := cluster.newNode(cluster.list.LocalNode())
+	cluster.hashStart = node.HashBegin
+	cluster.hashEnd = node.HashEnd
+	// And there aren't any others yet, so we're done.
+}
+
+// GetMembers returns the set of currently known cache members.
+func (cluster *Cluster) GetMembers() []*pb.Node {
+	// TODO(pebers): this is quadratic so would be bad on large clusters.
+	// We might also want to refresh the members in the background if this proves slow?
+	for _, m := range cluster.list.Members() {
+		cluster.newNode(m)
+	}
+	return cluster.nodes
+}
+
+// newNode constructs one of our canonical nodes from a memberlist.Node.
+// This includes allocating it hash space.
+func (cluster *Cluster) newNode(node *memberlist.Node) *pb.Node {
+	newNode := func(i int) *pb.Node {
+		return &pb.Node{
+			Name:      node.Name,
+			Address:   node.Addr.String() + ":" + string(node.Meta),
+			HashBegin: tools.HashPoint(i, cluster.size),
+			HashEnd:   tools.HashPoint(i+1, cluster.size),
+		}
+	}
+	for i, n := range cluster.nodes {
+		if n.Name == "" || n.Name == node.Name {
+			// Available slot. Or, if they identified as an existing node, they can take that space over.
+			if n.Name == node.Name {
+				log.Warning("Node %s / %s is taking over slot %d", node.Name, node.Addr, i)
+			} else {
+				log.Notice("Populating node %d: %s / %s", i, node.Name, node.Addr)
+			}
+			cluster.nodes[i] = newNode(i)
```

Comment:
Nope, it can't, but two concurrent requests might race and various things could then go wrong. I've added a `RWMutex` around `nodes` which should fix it.